flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -1066,6 +1066,8 @@ void markFinished(Map<String, Accumulator<?, ?>> userAccumulators, IOMetrics met
 			else if (current == CANCELING) {
 				// we sent a cancel call, and the task manager finished before it arrived. We
 				// will never get a CANCELED call back from the job manager
+				// release all partitions because partitions should only be kept if the execution reaches FINISHED
+				sendReleaseIntermediateResultPartitionsRpcCall();
```

Comment:
The release on cancel will be handled in [FLINK-13476](https://issues.apache.org/jira/browse/FLINK-13476)

Reply:
> At this point the PartitionTracker is not yet tracking these partitions (since we never officially reached a state FINISHED in the EG), hence the execution is sending these through separate RPC logic.

Yes, this is outdated and stems from an earlier version of the PR.

> Note that a similar issue can occur for pipelined partitions that are buffered in the producers side before a consumer was actually scheduled.
yes, partia...

Emphasis on "similar"; I'm referring to FLINK-13476 here which at the time did not exist yet.

Reply:
This code here only handles the state reconciliation in the case that an execution is finished while we're canceling.
If we go from RUNNING->CANCELING->CANCELED then the TM cleans up partitions. If an execution fails it is also cleaned up on the TM.